File: SORAL/python/testbed.py
#!/usr/bin/python

import soral

print('PySoral test based on tom.cpp')


def runTest():
    areas=4
    resources=1

    area = soral.doubleArray(areas)
    POA =  soral.doubleArray(areas)
    ESW =  soral.doubleArray(areas)
    speed =  soral.doubleArray(areas)

    # Configure the arrays with some test values
    for i in range(areas):
        area[i] = [28, 30, 14, 12][i]
        POA[i] = [0.1935,  0.2581,  0.2903,  0.2581][i]
        ESW[i] = 1
        speed[i] = 1

    availableHours = soral.doubleArray(resources)
    availableHours[0] = 40

    effectiveness = soral.Array2D(areas, resources)

    for resourceIdx in range(resources):
        for areaIdx in range(areas):
            value = ESW[areaIdx]*speed[areaIdx]/area[areaIdx]
            effectiveness.set(areaIdx, resourceIdx, value)
            # print "Effectiveness("+str(areaIdx)+","+str(resourceIdx)+" = "+str(value);


    theAllocation = soral.newCharnesCooper(resources, areas, effectiveness, availableHours, POA)

    print("The calculated allocation")
    printAssignments(theAllocation)

    print("PODs: ")
    for i in range(areas):
        print("  " +str(i) + ": " + str(theAllocation.getPOD(i)))

    print("Adjusted POAs: ")
    for i in range(areas):
        print("  " +str(i) + ": " + str(theAllocation.getNewPOC(i)))

    print("Segment POSs: ")
    for i in range(areas):
        print("  " +str(i) + ": " + str(theAllocation.getPOS(i)))

    print(" Cumulative POS: " + str(theAllocation.getTotalPOS()) )

    del theAllocation



def printAssignments( theAllocation ):
    activeItr = soral.ActiveAreasIterator(theAllocation)
    
    # While there are still areas with assignments
    while ( False == activeItr.atEnd() ):
        areaIndex = activeItr.getCurrentActiveAreaNum()
        area = soral.ActiveArea(areaIndex)

        resItr = soral.ResourceIterator(theAllocation, areaIndex)
    	
        while ( False == resItr.atEnd() ):
            resAssign = resItr.getResourceAssignment()
            resIndex = resAssign.getResourceNum()
            time = resAssign.getTime()
            
            print("  Area: " + str(areaIndex) + "  Resource: " + str(resIndex) + "  Time: " + str(time)	  )
            resItr.increment()
    
        activeItr.increment()


runTest()

File: SORAL/node/testbed.js
console=require("console");
soral=require("./build/Release/soral");


function runTest() {
    var i;
    
    console.log( "NodeSoral test based on tom.cpp" );
    
    var areas=4;
    var resources=1;
    
    var area = new soral.doubleArray(areas);
    var POA =  new soral.doubleArray(areas);
    var ESW =  new soral.doubleArray(areas);
    var speed = new soral.doubleArray(areas);
    
    // Configure the arrays with some test values
    for (i = 0; i < areas; i++) {
        area.setitem(i, [28, 30, 14, 12][i]);
        POA.setitem(i,  [0.1935,  0.2581,  0.2903,  0.2581][i]);
        ESW.setitem(i, 1);
        speed.setitem(i, 1);
    }

    var availableHours = new soral.doubleArray(resources);
    availableHours.setitem(0, 40);
    
    var effectiveness = new soral.Array2D(areas, resources);
    
    for (var resourceIdx  = 0; resourceIdx < resources; resourceIdx++) {
        for (var areaIdx = 0; areaIdx < areas; areaIdx++) {
            var value = ESW.getitem(areaIdx) * speed.getitem(areaIdx) / area.getitem(areaIdx);
            effectiveness.set(areaIdx, resourceIdx, value);
            // console.log("Effectiveness["+areaIdx+","+resourceIdx+"] = "+value + " // " + effectiveness.get(areaIdx, resourceIdx));
        }
    }
    
    var theAllocation = soral.newCharnesCooper(resources, areas, effectiveness, availableHours, POA);
    
    console.log("The calculated allocation");
    printAssignments(theAllocation);
    
    console.log("PODs: ");
    for (i = 0; i < areas; i++) { 
        console.log("  " +i + ": " + theAllocation.getPOD(i));
    }

    console.log("Adjusted POAs: ");
    for (i = 0; i < areas; i++) { 
        console.log("  " +i + ": " + theAllocation.getNewPOC(i));
    }

    console.log("Segment POSs: ");
    for (i = 0; i < areas; i++) { 
        console.log("  " +i + ": " + theAllocation.getPOS(i));
    }

    console.log(" Cumulative POS: " + theAllocation.getTotalPOS());
    
    delete theAllocation
	
}


function printAssignments( theAllocation ) {
    activeItr = new soral.ActiveAreasIterator(theAllocation);
    
    // While there are still areas with assignments
    while ( !activeItr.atEnd() ) {
        areaIndex = activeItr.getCurrentActiveAreaNum();
        area = new soral.ActiveArea(areaIndex);

    	resItr = new soral.ResourceIterator(theAllocation, areaIndex);
    	
    	while ( !resItr.atEnd() ) {
            resAssign = resItr.getResourceAssignment();
            resIndex = resAssign.getResourceNum();
            time = resAssign.getTime();
            
            console.log("  Area: " + areaIndex + "  Resource: " + resIndex + "  Time: " + time);	  
            resItr.increment();
        }
    	activeItr.increment();
	}
}






runTest();

